File: src/main/java/chapter4/conditional/SkipUntilExample.java
package chapter4.conditional;

import common.CommonUtils;
import common.Log;
import io.reactivex.Observable;

import java.util.concurrent.TimeUnit;


public class SkipUntilExample {

    public void marbleDiagram() {
        String[] data = {"1", "2", "3", "4", "5", "6"};

        Observable<String> source = Observable.fromArray(data)
                .zipWith(Observable.interval(100L, TimeUnit.MILLISECONDS), (val, notUsed) -> val)
                .skipUntil(Observable.timer(500L, TimeUnit.MILLISECONDS));

        source.subscribe(Log::it);
        CommonUtils.sleep(1000);
    }

    public static void main(String[] args) {
        SkipUntilExample demo = new SkipUntilExample();
        demo.marbleDiagram();
    }

}

File: src/main/kotlin/chapter4/conditional/SkipUntilExample_.kt
package chapter4.conditional

import common.CommonUtils_
import common.Log_
import io.reactivex.Observable
import io.reactivex.functions.BiFunction
import java.util.concurrent.TimeUnit

class SkipUntilExample_ {

    fun marbleDiagram() {
        val data: Array<String> = arrayOf("1", "2", "3", "4", "5", "6")

        val source = Observable.fromArray(*data)
                .zipWith(Observable.interval(100L, TimeUnit.MILLISECONDS), BiFunction { t1: String, _: Long -> t1 })
                .skipUntil(Observable.timer(500L, TimeUnit.MILLISECONDS))

        source.subscribe { Log_.it(it) }
        CommonUtils_.sleep(1000)
    }

}


fun main(args: Array<String>) {
    val demo = SkipUntilExample_()
    demo.marbleDiagram()
}
